include/libvfio-user.h

```diff
@@ -585,26 +585,41 @@ vfu_pci_set_class(vfu_ctx_t *vfu_ctx, uint8_t base, uint8_t sub, uint8_t pi);
 vfu_pci_config_space_t *
 vfu_pci_get_config_space(vfu_ctx_t *vfu_ctx);
 
-/* FIXME does it have to be packed as well? */
-typedef union {
-    struct msicap   msi;
-    struct msixcap  msix;
-    struct pmcap    pm;
-    struct pxcap    px;
-    struct vsc      vsc;
-} vfu_cap_t;
-
-//TODO: Support variable size capabilities.
+#define VFU_CAP_FLAG_EXTENDED (1 << 0)
+#define VFU_CAP_FLAG_READONLY (1 << 1)
 
 /**
- * Setup PCI capabilities.
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
+ *
+ * The contents of @data are copied in. It must start with either a struct
+ * pci_cap or a struct pci_extcap, with the ID field set (the 'next' field is
+ * ignored). The @flags values allows control over whether a client can write to
+ * the capability. Writes to the "id" or "next" fields are always disallowed.
+ *
+ * If @data is NULL, then it's expected that the region handler for
```

Comment:
I think it's cleaner to drop `cap_id` and always expect `data` to be non-null, therefore it must contain a valid PCI capability header where the capability ID can be found. We can add another flag, along with R/W access, to indicate that the capability must be handled by the user. At this point we can check that a callback has been set when the region is set up, and if not we can explicitly fail.

Reply:
We'd never use such data except during the initial add of the cap. That's a bit more confusing IMO: "Hey here's some data; but don't ever actually use it, I always want a callback instead."

But, on the other hand, it would mean we don't need the cap_id and size parameters.

Reply:
It might also be confusing to expect ID and size as a separate parameter while `data` must be correctly initialized (e.g. it must contain ID and size if it's the vendor-specific capability). It seems there's no ideal solution. If we decide to drop `cap_id` and `size`, then we can explain in the documentation that if the vendor-specific capability is added then `data` need only point to the the standed PCI capability header plus the size.

Another idea could be to have a separate function for adding vendor-specific capabilities, where we take ID, size, flags, and callback arguments separately. This will also simplify the function signature `vfu_pci_add_capability`, which will only be used for adding known capabilities.

Reply:
I thought about this some more, and I think you're right originally, we should require 'data' in all cases, and use a "CALLBACK" flag. Thinking about how a callback implementation would actually work, it really makes sense to copy in the data into place in the config space, then let the callback handle accesses, potentially just by memcpy()ing that location when needed.

Reply:
Makes sense. I suppose the user might decide to re-write what was just written by the libvfio-user before calling the callback?